Add an internal debug setting to enable Inline Predictions in all editable elements

rr-codes opened the pull request:
#### 2389cc0d2538658c4ef64ccf981cc2391f4b3a83
<pre>
Add an internal debug setting to enable Inline Predictions in all editable elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=259232">https://bugs.webkit.org/show_bug.cgi?id=259232</a>
rdar://112295387

Reviewed by Wenson Hsieh and Aditya Keerthi.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _updateTextInputTraits:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::allowsInlinePredictions const):

Canonical link: <a href="https://commits.webkit.org/266102@main">https://commits.webkit.org/266102@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c88676dcfbe80099de28c8e9248067bbad786f1d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12807 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13134 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13460 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14546 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12231 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15635 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13153 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14937 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12972 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13705 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10843 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14996 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10993 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11593 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18663 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/10889 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12068 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11762 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14963 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12120 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12211 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10123 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/12860 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11473 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3382 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3151 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15786 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13228 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12051 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3171 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5188,7 +5188,10 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
+    if (!m_page->preferences().inlinePredictionsEnabled() && !m_page->preferences().inlinePredictionsInAllEditableElementsEnabled())
+        return false;
+
+    return NSSpellChecker.isAutomaticInlineCompletionEnabled;
```
I also removed this selector check since it's no longer needed

pxlcoder commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -3186,6 +3186,19 @@ InlinePredictionsEnabled:
+  type: bool
+  status: internal
+  humanReadableName: "Inline Text Predictions"
+  humanReadableDescription: "Enable Inline Text Predictions in all contenteditable elements"
```
Would rewrite this and the PR title to refer to "editable elements" rather than "contenteditable elements", since it also covers `<input>` and `<textarea>`.

rr-codes replied:
Is it though? Because this check is also present
```
    if (!editorState.isContentEditable)
        return std::nullopt;
```

pxlcoder replied:
It does, since those elements have a `contenteditable` in their shadow DOM.

Though that's an implementation detail, whereas your flag is driving "user"-facing UI.

rr-codes replied:
ohhh, forgot they did that. Will change the  description / PR name

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5188,7 +5188,10 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
 bool WebViewImpl::allowsInlinePredictions() const
 {
-    return m_page->preferences().inlinePredictionsEnabled() && [NSSpellChecker respondsToSelector:@selector(isAutomaticInlineCompletionEnabled)] && NSSpellChecker.isAutomaticInlineCompletionEnabled;
+    if (!m_page->preferences().inlinePredictionsEnabled() && !m_page->preferences().inlinePredictionsInAllEditableElementsEnabled())
```
Now that I'm reading this, I think we should (in a separate PR) look into getting rid of the original `InlinePredictionsEnabled` from UnifiedWebPreferences. Since that preference isn't read in WebCore, or in the WebProcess, we could simply read off the `WKWebViewConfiguration`, without a need for a preference associated with the API.

Note how iOS already does this by checking `self.webView.configuration.allowsInlinePredictions`. The only problem is that `WebViewImpl` doesn't have easy access to `WKWebViewConfiguration` right now, which is why I suggest a separate PR.

whsieh replied:
> The only problem is that `WebViewImpl` doesn't have easy access to `WKWebViewConfiguration`

(Would it work to use `m_page->configuration()`, which returns the backing `API::Configuration`? Or alternately, `[m_view configuration]` should give the `WKWebViewConfiguration`, but at the cost of an extra copy of the configuration which is not ideal).

rr-codes committed:
Add an internal debug setting to enable Inline Predictions in all editable elements
https://bugs.webkit.org/show_bug.cgi?id=259232
rdar://112295387

Reviewed by Wenson Hsieh and Aditya Keerthi.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _updateTextInputTraits:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::allowsInlinePredictions const):

Canonical link: https://commits.webkit.org/266102@main
